src/mlpack/tests/decision_tree_regressor_test.cpp

```diff
@@ -0,0 +1,1047 @@
+          Approx(0.0).margin(1e-5));
+
+  REQUIRE(MSEGain::Evaluate<true>(labels, 0, weights) ==
+          Approx(0.0).margin(1e-5));
+}
+
+/**
+ * Making sure that MSE gain is evaluated correctly by doing calculation by
+ * hand.
+ */
+TEST_CASE("MSEGainHandCalculation", "[DecisionTreeRegressorTest]")
+{
+  arma::rowvec labels = {4., 2., 3., 4., 13., 6., 20., 8., 9., 10.};
+  arma::rowvec weights = {0.3, 0.3, 0.3, 0.3, 0.3, 0.7, 0.7, 0.7, 0.7, 0.7};
+
+  // Hand calculated gain values.
+  const double gain = -27.08999;
+  const double weightedGain = -27.53960;
+  REQUIRE(MSEGain::Evaluate<false>(labels, 0, weights) ==
+          Approx(gain).margin(1e-5));
+  REQUIRE(MSEGain::Evaluate<true>(labels, 0, weights) ==
+          Approx(weightedGain).margin(1e-5));
+}
+
+/**
+ * Make sure the MAD gain is zero when the labels are perfect.
+ */
+TEST_CASE("MADGainPerfectTest", "[DecisionTreeRegressorTest]")
+{
+  arma::rowvec weights(10, arma::fill::ones);
+  arma::rowvec labels;
+  labels.ones(10);
+
+  REQUIRE(MADGain::Evaluate<false>(labels, 0, weights) ==
+          Approx(0.0).margin(1e-5));
+}
+
+/**
+ * Make sure that when mean of labels is zero, MAD_gain = mean of
+ * absolute values of the distribution.
+ */
+TEST_CASE("MADGainNormalTest", "[DecisionTreeRegressorTest")
+{
+  arma::rowvec weights(10, arma::fill::ones);
+  arma::rowvec labels = { 1, 2, 3, 4, 5, -1, -2, -3, -4, -5 }; // Mean = 0.
+
+  // Theoretical gain.
+  double theoreticalGain = 0.0;
+  for (size_t i = 0; i < labels.n_elem; ++i)
+    theoreticalGain -= std::abs(labels[i]);
+  theoreticalGain /= (double) labels.n_elem;
+
+  // Calculated gain.
+  const double calculatedGain = MADGain::Evaluate<false>(labels, 0, weights);
+
+  REQUIRE(calculatedGain == Approx(theoreticalGain).margin(1e-5));
+}
+
+/**
+ * The MAD gain of an empty vector is 0.
+ */
+TEST_CASE("MADGainEmptyTest", "[DecisionTreeRegressorTest]")
+{
+  arma::rowvec weights = arma::ones<arma::rowvec>(10);
+  arma::rowvec labels;
+  REQUIRE(MADGain::Evaluate<false>(labels, 0, weights) ==
+          Approx(0.0).margin(1e-5));
+
+  REQUIRE(MADGain::Evaluate<true>(labels, 0, weights) ==
+          Approx(0.0).margin(1e-5));
+}
+
+/**
+ * Making sure that MAD gain is evaluated correctly by doing calculation by
+ * hand.
+ */
+TEST_CASE("MADGainHandCalculation", "[DecisionTreeRegressorTest]")
+{
+  arma::rowvec labels = {4., 2., 3., 4., 13., 6., 20., 8., 9., 10.};
+  arma::rowvec weights = {0.3, 0.3, 0.3, 0.3, 0.3, 0.7, 0.7, 0.7, 0.7, 0.7};
+
+  // Hand calculated gain values.
+  const double gain = -4.1;
+  const double weightedGain = -3.8592;
+  REQUIRE(MADGain::Evaluate<false>(labels, 0, weights) ==
+          Approx(gain).margin(1e-5));
+  REQUIRE(MADGain::Evaluate<true>(labels, 0, weights) ==
+          Approx(weightedGain).margin(1e-5));
+}
+
+/**
+ * Check that AllCategoricalSplit will split when the split is obviously
+ * better.
+ */
+TEST_CASE("AllCategoricalSplitSimpleSplitTest_", "[DecisionTreeRegressorTest]")
```

Comment:
I was wondering when you were gonna ask me this. :-)
Actually, these tests are also in `decision_tree_test.cpp`. So, in "catch2", we can have only one test with a particular name. It won't accept even though they are in different test suits. So, I added an underscore to them.

Reply:
Ah, understood.  I'm ok with that, but you could also just make them all `...RegressionTest` or `...ClassificationTest` instead of `...Test_`---up to you. :+1:

Reply:
I think it will be redundant because the name of the test suite is "DecisionTreeRegressor".

Reply:
Yeah, I can see what you mean.  No worries leaving it as-is. :+1: